File: trunk/D2Packets/BnetClientPackets.cs
using System;
using D2Data;
using D2Packets;
using ETUtils;

namespace BnetClient
{
    /// <summary>
	/// Base class for Battle.net Client Packets
	/// </summary>
	public class BCPacket : D2Packet
	{
		public readonly BnetClientPacket PacketID;

		public BCPacket(byte[] data) : base(data, Origin.BattleNetClient)
		{
			this.PacketID = (BnetClientPacket) data[1];
		}
	}

    /// <summary>
    /// BC 0x00
    /// </summary>
    public class KeepAlive : BCPacket
    {
        public KeepAlive(byte[] data)
            : base(data)
        {
        }
    }

    public class LeaveChat : BCPacket
    {
        public LeaveChat(byte[] data)
            : base(data)
        {

        }
    }


    /// <summary>
    /// BC Packet 0x15
    /// </summary>
    public class AdInfoRequest : BCPacket
    {
        public readonly  BattleNetClient Client;
        public readonly uint Id;
        public readonly BattleNetPlatform Platform;
        public readonly DateTime Timestamp;

        public AdInfoRequest(byte[] data)
            : base(data)
        {
            this.Platform = (BattleNetPlatform)BitConverter.ToUInt32(data, 3);
            this.Client = (BattleNetClient)BitConverter.ToUInt32(data, 7);
            this.Id = BitConverter.ToUInt32(data, 11);
            this.Timestamp = TimeUtils.ParseUnixTimeUtc(BitConverter.ToUInt32(data, 15));
        }

    }

    [Flags]
    public enum StartGameFlags
    {
        Public,
        Private
    }

 /// <summary>
 /// BC 0x1C
 /// </summary>
    public class StartGame : BCPacket
    {
        public readonly StartGameFlags flags;
        public readonly string name;
        public readonly string password;
        public readonly string statString;

        public StartGame(byte[] data)
            : base(data)
        {
            this.flags = (StartGameFlags)BitConverter.ToUInt32(data, 3);
            this.name = ByteConverter.GetNullString(data, 0x17);
            if ((this.flags & StartGameFlags.Private) == StartGameFlags.Private)
            {
                this.password = ByteConverter.GetNullString(data, 0x18 + this.name.Length);
            }
            int offset = (0x19 + this.name.Length) + ((this.password == null) ? 0 : this.password.Length);
            if (data.Length > (offset + 1))
            {
                this.statString = ByteConverter.GetNullString(data, offset);
            }
        }
    }


	/// <summary>
	/// BC Packet 0x1F - Leave Game - Notifies Battle.net user has left a game.
	/// </summary>
	public class LeaveGame : BCPacket
	{
		public LeaveGame(byte[] data) : base (data)
		{
		}
	}

    /// <summary>
    /// BC 0x21
    /// </summary>
    public class DisplayAd : BCPacket
    {
        public readonly BattleNetClient client;
        public readonly string filename;
        public readonly uint id;
        public readonly BattleNetPlatform platform;
        public readonly  string url;

        public DisplayAd(byte[] data)
            : base(data)
        {
            this.platform = (BattleNetPlatform)BitConverter.ToUInt32(data, 3);
            this.client = (BattleNetClient)BitConverter.ToUInt32(data, 7);
            this.id = BitConverter.ToUInt32(data, 11);
            if (data[15] != 0)
            {
                this.filename = ByteConverter.GetNullString(data, 15);
            }
            int index = 0x10 + ((this.filename == null) ? 0 : this.filename.Length);
            if (data[index] != 0)
            {
                this.url = ByteConverter.GetNullString(data, index);
            }
        }

    }

    /// <summary>
    /// BC 0x22
    /// </summary>
    public class NotifyJoin : BCPacket
    {
        public readonly BattleNetClient client;
        public readonly string name;
        public readonly string password;
        public readonly uint version;

        public NotifyJoin(byte[] data)
            : base(data)
        {
            this.client = (BattleNetClient)BitConverter.ToUInt32(data, 3);
            this.version = BitConverter.ToUInt32(data, 7);
            this.name = ByteConverter.GetNullString(data, 11);
            this.password = ByteConverter.GetNullString(data, 12 + this.name.Length);
        }
    }

    /// <summary>
    /// BC 0x25
    /// </summary>
    public class BnetPong : BCPacket
    {
        public readonly  uint timestamp;

        public BnetPong(byte[] data)
            : base(data)
        {
            this.timestamp = BitConverter.ToUInt32(data, 3);
        }
    }

    /// <summary>
    /// BC 0x33
    /// </summary>
    public class FileTimeRequest : BCPacket
    {
        public readonly string filename;
        public readonly uint requestID;
        public readonly uint unknown;


        public FileTimeRequest(byte[] data)
            : base(data)
        {
            this.requestID = BitConverter.ToUInt32(data, 3);
            this.unknown = BitConverter.ToUInt32(data, 7);
            this.filename = ByteConverter.GetNullString(data, 11);
        }
    }

    /// <summary>
    /// BC Packet 0x3A - BnetLogonRequest
    /// </summary>
    public class BnetLogonRequest : BCPacket
    {
        public readonly uint clientToken;
        public readonly uint serverToken;
        public readonly string username;

        public BnetLogonRequest(byte[] data)
            : base(data)
        {
            this.clientToken = BitConverter.ToUInt32(data, 3);
            this.serverToken = BitConverter.ToUInt32(data, 7);
            this.username = ByteConverter.GetNullString(data, 0x1f);
        }


    }

	/// <summary>
	/// BC Packet 0x3E - Realm Logon Request - Request Realm server information to establish a connection.
	/// </summary>
	public class RealmLogonRequest : BCPacket
	{
		public readonly string Realm;
		public readonly uint Cookie;

		public RealmLogonRequest(byte[] data) : base (data)
		{
			this.Cookie = BitConverter.ToUInt32(data, 3);
			// 5 * DWORD; hashed password ("password")
			this.Realm = ByteConverter.GetNullString(data, 27);
		}
	}

    /// <summary>
    /// BC 0x40
    /// </summary>
    public class QueryRealms : BCPacket
    {
        public QueryRealms(byte[] data)
            : base(data)
        {
        }
    }

    /// <summary>
    /// BC 0x46
    /// </summary>
    public class NewsInfoRequest : BCPacket
    {
        public readonly DateTime since;

        public NewsInfoRequest(byte[] data)
            : base(data)
        {
            this.since = TimeUtils.ParseUnixTimeUtc(BitConverter.ToUInt32(data, 3));
        }
    }

    /// <summary>
    /// BC 0x4B
    /// </summary>
    public class ExtraWorkResponse : BCPacket
    {
        public readonly  int client;
        public readonly int resultLength;

        public byte[] ResultData
        {
            get
            {
                byte[] destinationArray = new byte[this.resultLength];
                Array.Copy(base.Data, 7, destinationArray, 0, this.resultLength);
                return destinationArray;
            }
        }

        public ExtraWorkResponse(byte[] data)
            : base(data)
        {
            this.client = BitConverter.ToUInt16(data, 3);
            this.resultLength = BitConverter.ToUInt16(data, 5);
        }
    }

    /// <summary>
    /// BC Packet 0x50 - Bnet Connection Request - First packet sent to bnet to establish a connection.
    /// </summary>
    public class BnetConnectionRequest : BCPacket
    {
        public readonly BattleNetClient client;
        public readonly string countryAbbreviation;
        public readonly string countryName;
        public readonly static uint CurrentD2LoDVersion;
        public readonly static uint CurrentD2Version;
        public readonly uint language;
        public readonly uint languageID;
        public readonly uint localeID;
        public readonly System.Net.IPAddress localIP;
        public readonly BattleNetPlatform platform;
        public readonly uint protocol;
        public readonly uint timeZoneBias;
        public readonly uint version;

        public BnetConnectionRequest(byte[] data)
            : base(data)
        {
            this.protocol = BitConverter.ToUInt32(data, 3);
            this.platform = (BattleNetPlatform)BitConverter.ToUInt32(data, 7);
            this.client = (BattleNetClient)BitConverter.ToUInt32(data, 11);
            this.version = BitConverter.ToUInt32(data, 15);
            this.language = BitConverter.ToUInt32(data, 0x13);
            this.localIP = new System.Net.IPAddress((long)BitConverter.ToUInt32(data, 0x17));
            this.timeZoneBias = BitConverter.ToUInt32(data, 0x1b);
            this.localeID = BitConverter.ToUInt32(data, 0x1f);
            this.languageID = BitConverter.ToUInt32(data, 0x23);
            this.countryAbbreviation = ByteConverter.GetNullString(data, 0x27);
            this.countryName = ByteConverter.GetNullString(data, 40 + this.countryAbbreviation.Length);
        }
    }

    public struct CDKeyInfo
    {
        uint Length;
        uint ProductValue;
        uint PublicValue;
        uint Unknown;       // 0
        byte[] Hash;

        public CDKeyInfo(byte[] data, int offset)
        {
            this.Length = BitConverter.ToUInt32(data, offset);
            this.ProductValue = BitConverter.ToUInt32(data, offset + 4);
            this.PublicValue = BitConverter.ToUInt32(data, offset + 8);
            this.Unknown = BitConverter.ToUInt32(data, offset + 12);
            this.Hash = new byte[20];
            Array.Copy(data, offset + 16, this.Hash, 0, 20);
        }
    }

    /// <summary>
    /// BC Packet 0x51 - Bnet Auth Request - Second packet sent to bnet containing CDKey and other auth info.
    /// </summary>
    public class BnetAuthRequest : BCPacket
    {
        public readonly uint ClientToken;
        public readonly uint GameVersion;
        public readonly uint GameHash;
        public readonly uint KeyCount;      // 1 for classic, 2 for LOD...
        public readonly bool UseSpawn;      // Always false for Diablo...
        public readonly CDKeyInfo[] Keys;
        public readonly string GameInfo;
        public readonly string OwnerName;

        public BnetAuthRequest(byte[] data) : base(data)
        {
            this.ClientToken = BitConverter.ToUInt32(data, 3);
            this.GameVersion = BitConverter.ToUInt32(data, 7);
            this.GameHash = BitConverter.ToUInt32(data, 11);
            this.KeyCount = BitConverter.ToUInt32(data, 15);
            this.UseSpawn = BitConverter.ToUInt32(data, 19) == 1 ? true : false;
            this.Keys = new CDKeyInfo[this.KeyCount];
            int offset = 23;
            for (int i = 0; i < this.KeyCount; i++)
            {
                this.Keys[i] = new CDKeyInfo(data, offset);
                offset += 36;
            }
            this.GameInfo = ByteConverter.GetNullString(data, offset);
            this.OwnerName = ByteConverter.GetNullString(data, offset + this.GameInfo.Length + 1);
        }
    }

    /// <summary>
    /// BC 0x0a
    /// </summary>
    public class EnterChatRequest : BCPacket
    {
        public readonly string name;
        public readonly string realm;

        public EnterChatRequest(byte[] data)
            : base(data)
        {
            this.name = ByteConverter.GetNullString(data, 3);
            this.realm = ByteConverter.GetString(data, 4 + this.name.Length, -1, 0x2c);
        }
    }
    
    /// <summary>
    /// BC 0x0B
    /// </summary>
    public class ChannelListRequest : BCPacket
    {
        public readonly  BattleNetClient client;

        public ChannelListRequest(byte[] data)
            : base(data)
        {
            this.client = (BattleNetClient)BitConverter.ToUInt32(data, 3);
        }

    }

    public enum JoinChannelFlags
    {
        AutoJoin = 5,
        Create = 2,
        NormalJoin = 0
    }

    /// <summary>
    /// BC 0x0C
    /// </summary>
    public class JoinChannel : BCPacket
    {
        public readonly JoinChannelFlags flags;
        public readonly string name;

        public JoinChannel(byte[] data)
            : base(data)
        {
            this.flags = (JoinChannelFlags)BitConverter.ToUInt32(data, 3);
            this.name = ByteConverter.GetNullString(data, 7);
        }
    }

    /// <summary>
    /// BC 0x0E
    /// </summary>
    public class ChatCommand : BCPacket
    {
        public readonly string message;

        public ChatCommand(byte[] data)
            : base(data)
        {
            this.message = ByteConverter.GetNullString(data, 3);
        }

 


    }

}

File: trunk/D2PacketsVB/D2PacketsVB/Bnet/Client/BnetClientPackets.vb
Imports System
Imports D2Data

Imports ETUtils
Imports D2Packets
Imports System.Globalization

Namespace BnetClient

    Public Class BcPacket
        Inherits D2Packet

        ' Fields
        Public ReadOnly Packetid As Byte

        ' Methods
        Public Sub New(ByVal Id As Byte)
            Packetid = Id
        End Sub
        Public Sub New(ByVal Data() As Byte)

            Me.Packetid = Data(1)
            Dim PacketData(Data.Length - 4) As Byte
            Buffer.BlockCopy(Data, 4, PacketData, 0, Data.Length - 4)

            MyBase.InsertByteArray(PacketData)

        End Sub

        Public Overrides ReadOnly Property Data() As Byte()
            Get
                Dim dst As Byte() = New Byte(MyBase.Count + 3) {}
                Dim dat As Byte() = MyBase.GetData
                dst(0) = &HFF
                dst(1) = Packetid
                Dim bytes As Byte() = BitConverter.GetBytes(CUShort((MyBase.Count + 4 And &HFFFF)))
                dst(2) = bytes(0)
                dst(3) = bytes(1)
                Buffer.BlockCopy(dat, 0, dst, 4, dat.Length)
                Return dst
            End Get
        End Property

    End Class

    Public Class AdInfoRequest
        Inherits BcPacket

        ' Fields
        Public ReadOnly Client As BattleNetClient
        Public ReadOnly Id As UInteger
        Public ReadOnly Platform As BattleNetPlatform
        Public ReadOnly Timestamp As DateTime

        Public Sub New(ByVal Id As UInteger, ByVal TimeStamp As UInteger, Optional ByVal Expansion As Boolean = True)
            MyBase.new(BnetClientPacket.AdInfoRequest)

            InsertDwordString("IX86")     ' (DWORD)  Platform Code

            If Expansion Then
                InsertDwordString("D2XP") ' (DWORD)  Product Code
                Me.Client = BattleNetClient.Diablo2LoD
            Else
                InsertDwordString("D2DV") ' (DWORD)  Product Code
                Me.Client = BattleNetClient.Diablo2
            End If

            InsertUInt32(Id)
            InsertUInt32(TimeStamp)

            Me.Platform = BattleNetPlatform.Ix86

            Me.Id = Id
            Me.Timestamp = DateTime.FromBinary(TimeStamp)


        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.Platform = DirectCast(BitConverter.ToUInt32(data, 4), BattleNetPlatform)
            Me.Client = DirectCast(BitConverter.ToUInt32(data, 8), BattleNetClient)
            Me.Id = BitConverter.ToUInt32(data, 12)
            Me.Timestamp = TimeUtils.ParseUnixTimeUtc(BitConverter.ToUInt32(data, 16))
        End Sub

    End Class


    Public Class BnetAuthRequest
        Inherits BcPacket

        ' Fields
        Public ReadOnly clientToken As UInt32
        Public ReadOnly gameHash As UInt32
        Public ReadOnly gameInfo As String
        Public ReadOnly gameVersion As UInt32
        Public ReadOnly keyCount As UInt32
        Public ReadOnly keys As Struct.CDKeyInfo()
        Public ReadOnly ownerName As String
        Public ReadOnly useSpawn As Boolean

        ' Methods
        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.clientToken = BitConverter.ToUInt32(data, 3)
            Me.gameVersion = BitConverter.ToUInt32(data, 7)
            Me.gameHash = BitConverter.ToUInt32(data, 11)
            Me.keyCount = BitConverter.ToUInt32(data, 15)
            Me.useSpawn = (BitConverter.ToUInt32(data, &H13) = 1)
            Me.keys = New Struct.CDKeyInfo(Me.keyCount - 1) {}
            Dim offset As Integer = &H17
            Dim i As Integer
            For i = 0 To Me.keyCount - 1
                Me.keys(i) = New Struct.CDKeyInfo(data, offset)
                offset = (offset + &H24)
            Next i
            Me.gameInfo = ByteConverter.GetNullString(data, offset)
            Me.ownerName = ByteConverter.GetNullString(data, ((offset + Me.gameInfo.Length) + 1))
        End Sub

    End Class

    Public Class BnetConnectionRequest
        Inherits BcPacket

        Public ReadOnly client As BattleNetClient
        Public ReadOnly countryAbbreviation As String
        Public ReadOnly countryName As String
        Public Shared ReadOnly CurrentD2LoDVersion As UInt32 = 11
        Public Shared ReadOnly CurrentD2Version As UInt32 = 11
        Public ReadOnly language As UInt32
        Public ReadOnly languageID As UInt32
        Public ReadOnly localeID As UInt32
        Public ReadOnly localIP As Net.IPAddress
        Public ReadOnly platform As BattleNetPlatform
        Public ReadOnly protocol As UInt32
        Public ReadOnly timeZoneBias As UInt32
        Public ReadOnly version As UInt32


        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.protocol = BitConverter.ToUInt32(data, 3)
            Me.platform = DirectCast(BitConverter.ToUInt32(data, 7), BattleNetPlatform)
            Me.client = DirectCast(BitConverter.ToUInt32(data, 11), BattleNetClient)
            Me.version = BitConverter.ToUInt32(data, 15)
            Me.language = BitConverter.ToUInt32(data, &H13)
            Me.localIP = New Net.IPAddress(CLng(BitConverter.ToUInt32(data, &H17)))
            Me.timeZoneBias = BitConverter.ToUInt32(data, &H1B)
            Me.localeID = BitConverter.ToUInt32(data, &H1F)
            Me.languageID = BitConverter.ToUInt32(data, &H23)
            Me.countryAbbreviation = ByteConverter.GetNullString(data, &H27)
            Me.countryName = ByteConverter.GetNullString(data, (40 + Me.countryAbbreviation.Length))
        End Sub

        ' Methods
        Public Sub New(Optional ByVal Expansion As Boolean = True, Optional ByVal VersionByte As Integer = &HC)
            MyBase.New(BnetClientPacket.BnetConnectionRequest)

            InsertInt32(&H0)                                                 ' (DWORD)  Protocol ID
            InsertDwordString("IX86")                                        ' (DWORD)  Platform Code

            If Expansion Then
                InsertDwordString("D2XP")                                    ' (DWORD)  Product Code
            Else
                InsertDwordString("D2DV")                                    ' (DWORD)  Product Code
            End If

            InsertInt32(VersionByte)                                         ' (DWORD)  Version Byte
            InsertInt32(CultureInfo.CurrentUICulture.LCID)                   ' (DWORD)  Product Language
            InsertInt32(&H100007F)                                           ' (DWORD)   Local IP

            InsertInt32(DateTime.UtcNow.Subtract(DateTime.Now).TotalMinutes) ' (DWORD)  Timezone Bias

            InsertInt32(CultureInfo.CurrentUICulture.LCID)                   ' (DWORD)  Locale ID
            InsertInt32(CultureInfo.CurrentUICulture.LCID)                   ' (DWORD)  Language ID
            InsertCString(RegionInfo.CurrentRegion.ThreeLetterISORegionName) ' (STRING) Country Abbreviation
            InsertCString(RegionInfo.CurrentRegion.DisplayName)              ' (STRING) Country name

        End Sub

    End Class

    Public Class BnetLogonRequest
        Inherits BcPacket

        Public ReadOnly clientToken As UInt32
        Public ReadOnly serverToken As UInt32
        Public ReadOnly username As String

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)

            Me.clientToken = BitConverter.ToUInt32(data, 4)
            Me.serverToken = BitConverter.ToUInt32(data, 8)
            Me.username = ByteConverter.GetNullString(data, 32)

        End Sub

    End Class

    Public Class BnetPong
        Inherits BcPacket

        Public ReadOnly timestamp As UInt32

        ' Methods
        Public Sub New(ByVal TimeStamp As UInteger)

            MyBase.New(BnetClientPacket.BnetPong)
            InsertUInt32(TimeStamp)

            Me.timestamp = TimeStamp
        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.timestamp = BitConverter.ToUInt32(data, 4)
        End Sub


    End Class

    Public Class ChannelListRequest
        Inherits BcPacket

        Public ReadOnly client As BattleNetClient

        ' Methods
        Public Sub New()
            MyBase.New(BnetClientPacket.ChannelListRequest)
            InsertInt32(0)
            client = BattleNetClient.Unknown
        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.client = DirectCast(BitConverter.ToUInt32(data, 4), BattleNetClient)
        End Sub

    End Class

    Public Class ChatCommand
        Inherits BcPacket

        Public ReadOnly Message As String

        ' Methods
        Public Sub New(ByVal Message As String)
            MyBase.New(BnetClientPacket.ChatCommand)
            InsertCString(Message, System.Text.Encoding.UTF8)
        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.Message = ByteConverter.GetNullString(data, 4)
        End Sub


    End Class

    Public Class DisplayAd
        Inherits BcPacket

        Public ReadOnly Client As BattleNetClient
        Public ReadOnly Filename As String
        Public ReadOnly ID As UInt32
        Public ReadOnly Platform As BattleNetPlatform
        Public ReadOnly URL As String

        ' Methods
        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.Platform = DirectCast(BitConverter.ToUInt32(data, 4), BattleNetPlatform)
            Me.Client = DirectCast(BitConverter.ToUInt32(data, 8), BattleNetClient)
            Me.ID = BitConverter.ToUInt32(data, 12)
            If (data(15) <> 0) Then
                Me.Filename = ByteConverter.GetNullString(data, 16)
            End If
            Dim index As Integer = (17 + IIf((Me.Filename Is Nothing), 0, Me.Filename.Length))
            If (data(index) <> 0) Then
                Me.url = ByteConverter.GetNullString(data, index)
            End If
        End Sub


    End Class

    Public Class EnterChatRequest
        Inherits BcPacket

        Public ReadOnly Name As String
        Public ReadOnly Realm As String

        Public Sub New(ByVal Username As String)
            MyBase.New(BnetClientPacket.EnterChatRequest)

            InsertCString(Username)
            InsertByte(0)

        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.Name = ByteConverter.GetNullString(data, 4)
            Me.Realm = ByteConverter.GetString(data, (5 + Me.Name.Length), -1, 44)
        End Sub

    End Class

    Public Class ExtraWorkResponse
        Inherits BcPacket

        Public ReadOnly client As Integer
        Public ReadOnly resultLength As Integer

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.client = BitConverter.ToUInt16(data, 4)
            Me.resultLength = BitConverter.ToUInt16(data, 6)
        End Sub

    End Class

    Public Class FileTimeRequest
        Inherits BcPacket

        Public ReadOnly filename As String
        Public ReadOnly requestID As UInt32

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.requestID = BitConverter.ToUInt32(data, 4)
            Me.filename = ByteConverter.GetNullString(data, 12)
        End Sub


    End Class


    Public Enum JoinChannelFlags
        ' Fields
        AutoJoin = 5
        Create = 2
        NormalJoin = 0
    End Enum
    Public Class JoinChannel
        Inherits BcPacket

        Protected Flags As JoinChannelFlags
        Protected Name As String

        Public Sub New(ByVal Name As String, Optional ByVal FirstJoin As Boolean = False)
            MyBase.New(BnetClientPacket.JoinChannel)

            If FirstJoin Then
                InsertInt32(&H5)
                InsertCString("Diablo II")
            Else
                InsertInt32(&H2)
                InsertCString(Name)
            End If

        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.Flags = BitConverter.ToInt32(data, 4)
            Me.Name = ByteConverter.GetNullString(data, 8)
        End Sub


    End Class

    Public Class KeepAlive
        Inherits BcPacket

        ' Methods
        Public Sub New()
            MyBase.New(BnetClientPacket.KeepAlive)
        End Sub
        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
        End Sub

    End Class

    Public Class LeaveChat
        Inherits BcPacket

        Public Sub New()
            MyBase.New(BnetClientPacket.LeaveChat)
        End Sub
        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
        End Sub

    End Class

    Public Class LeaveGame
        Inherits BcPacket

        Public Sub New()
            MyBase.New(BnetClientPacket.LeaveGame)
        End Sub
        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
        End Sub

    End Class

    Public Class NewsInfoRequest
        Inherits BcPacket
        Public ReadOnly Since As DateTime

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.Since = TimeUtils.ParseUnixTimeUtc(BitConverter.ToUInt32(data, 4))
        End Sub


    End Class

    Public Class NotifyJoin
        Inherits BcPacket


        Public ReadOnly client As BattleNetClient
        Public ReadOnly name As String
        Public ReadOnly password As String
        Public ReadOnly version As UInt32

        Sub New(ByVal Client As D2Data.BattleNetClient, ByVal GameName As String, Optional ByVal GamePassword As String = "", Optional ByVal Version As Integer = &HC)
            MyBase.New(BnetClientPacket.NotifyJoin)

            InsertUInt32(Client)
            InsertUInt32(Version)
            InsertCString(GameName)
            InsertCString(GamePassword)

        End Sub

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.client = DirectCast(BitConverter.ToUInt32(data, 4), BattleNetClient)
            Me.version = BitConverter.ToUInt32(data, 8)
            Me.name = ByteConverter.GetNullString(data, 12)
            Me.password = ByteConverter.GetNullString(data, (13 + Me.name.Length))
        End Sub


    End Class

    Public Class QueryRealms
        Inherits BcPacket

        ' Methods
        Public Sub New()
            MyBase.New(BnetClientPacket.QueryRealms)
        End Sub
        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
        End Sub
    End Class

    Public Class RealmLogonRequest
        Inherits BcPacket

        Public ReadOnly cookie As UInt32
        Public ReadOnly realm As String

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.cookie = BitConverter.ToUInt32(data, 4)
            Me.realm = ByteConverter.GetNullString(data, 28)
        End Sub


    End Class

    Public Class StartGame
        Inherits BcPacket

        Public ReadOnly flags As StartGameFlags
        Public ReadOnly name As String
        Public ReadOnly password As String
        Public ReadOnly statString As String

        Public Sub New(ByVal data As Byte())
            MyBase.New(data)
            Me.flags = BitConverter.ToInt32(data, 4)
            Me.name = ByteConverter.GetNullString(data, 24)
            If ((Me.flags And StartGameFlags.Private) = StartGameFlags.Private) Then
                Me.password = ByteConverter.GetNullString(data, (25 + Me.name.Length))
            End If
            Dim offset As Integer = ((26 + Me.name.Length) + IIf((Me.password Is Nothing), 0, Me.password.Length))
            If (data.Length > (offset + 1)) Then
                Me.statString = ByteConverter.GetNullString(data, offset)
            End If
        End Sub

    End Class

    <Flags()> _
    Public Enum StartGameFlags
        [Public]
        [Private]
    End Enum

End Namespace



